### PR TITLE
Fix OAuth metadata discovery for path-based server URLs

### LINKED ISSUE #697: OAuth metadata discovery fails for servers with path components (RFC 8414/9728 compliance)
## Summary

OAuth metadata discovery fails for MCP servers that use path components in their URLs or have separate domains for MCP and OAuth services. This affects servers like Smithery that use:

- MCP Server: `https://server.smithery.ai/googledrive`
- Auth Server: `https://auth.smithery.ai/googledrive`

### 1. Base URL Path Stripping

In `client/transport/streamable_http.go` (lines 163-168), the base URL construction strips the path:

```go
if smc.oauthHandler != nil {
    baseURL := fmt.Sprintf("%s://%s", parsedURL.Scheme, parsedURL.Host)
    smc.oauthHandler.SetBaseURL(baseURL)
}
```

For `https://server.smithery.ai/googledrive`, this sets baseURL to `https://server.smithery.ai`, causing:

- PRM discovery at `https://server.smithery.ai/.well-known/oauth-protected-resource` (wrong)
- Should be: `https://server.smithery.ai/.well-known/oauth-protected-resource/googledrive` (RFC 9728)

### 2. RFC 8414 Path Insertion Missing

The current code only tries one metadata URL format:

```go
authServerURL + "/.well-known/oauth-authorization-server"
```

RFC 8414 Section 3.1 specifies that when the issuer URL contains a path, the well-known path should be **inserted between host and path**:

- Input: `https://auth.smithery.ai/googledrive`
- Current: `https://auth.smithery.ai/googledrive/.well-known/oauth-authorization-server` ❌
- RFC 8414: `https://auth.smithery.ai/.well-known/oauth-authorization-server/googledrive` ✅

### 3. No WWW-Authenticate Header Parsing

RFC 9728 Section 3 specifies that servers should advertise their Protected Resource Metadata URL via the `WWW-Authenticate` header:

```
WWW-Authenticate: Bearer error="invalid_request", resource_metadata="https://server.smithery.ai/.well-known/oauth-protected-resource/googledrive"
```

The current implementation doesn't parse this header to extract the `resource_metadata` URL.

## Impact

- **DCR fails with 404** because the registration endpoint URL is constructed incorrectly
- **Authorization URL has empty client_id** because DCR failed
- **Affects all servers** using path-based URLs or separate auth domains (Smithery, potentially others)

### 1. RFC 8414 Compliant URL Construction

```go
// BuildRFC8414MetadataURLs constructs OAuth Authorization Server Metadata URLs per RFC 8414.
func BuildRFC8414MetadataURLs(authServerURL string) []string {
    u, err := url.Parse(authServerURL)
    if err != nil {
        return []string{authServerURL + "/.well-known/oauth-authorization-server"}
    }

    path := strings.TrimSuffix(u.Path, "/")
    baseURL := fmt.Sprintf("%s://%s", u.Scheme, u.Host)

    var urls []string
    if path == "" || path == "/" {
        urls = append(urls, baseURL+"/.well-known/oauth-authorization-server")
    } else {
        // RFC 8414: insert .well-known between host and path
        rfc8414URL := baseURL + "/.well-known/oauth-authorization-server" + path
        urls = append(urls, rfc8414URL)
        // Legacy fallback
        legacyURL := strings.TrimSuffix(authServerURL, "/") + "/.well-known/oauth-authorization-server"
        urls = append(urls, legacyURL)
        // Base URL fallback (for servers like Cloudflare)
        urls = append(urls, baseURL+"/.well-known/oauth-authorization-server")
    }
    return urls
}
```

### 2. WWW-Authenticate Header Parsing

```go
// ExtractResourceMetadataURL parses WWW-Authenticate header for resource_metadata URL
func ExtractResourceMetadataURL(wwwAuthHeader string) string {
    if !strings.Contains(wwwAuthHeader, "resource_metadata") {
        return ""
    }
    parts := strings.Split(wwwAuthHeader, `resource_metadata="`)
    if len(parts) < 2 {
        return ""
    }
    endIdx := strings.Index(parts[1], `"`)
    if endIdx == -1 {
        return ""
    }
    return parts[1][:endIdx]
}
```

### 3. Auth Server Discovery from PRM

When the MCP server and OAuth server are on different domains, discover the auth server from Protected Resource Metadata before fetching OAuth metadata.

## Related Issue

We encountered this while implementing OAuth support in MCPProxy:

- **Issue**: https://github.com/smart-mcp-proxy/mcpproxy-go/issues/251
- **Fix**: https://github.com/smart-mcp-proxy/mcpproxy-go/pull/262

Our workaround implements the discovery logic in mcpproxy, but this would benefit all mcp-go users.

## Test Case

Smithery's googledrive server can be used to verify the fix:

```bash
# Protected Resource Metadata (RFC 9728)
curl -s "https://server.smithery.ai/.well-known/oauth-protected-resource/googledrive" | jq .
# Returns: {"authorization_servers": ["https://auth.smithery.ai/googledrive"], ...}

# OAuth Authorization Server Metadata (RFC 8414)
curl -s "https://auth.smithery.ai/.well-known/oauth-authorization-server/googledrive" | jq .
# Returns: {"issuer": "...", "registration_endpoint": "...", ...}

# DCR endpoint
curl -s -X POST "https://auth.smithery.ai/googledrive/register" \
  -H "Content-Type: application/json" \
  -d '{"redirect_uris": ["http://127.0.0.1:8080/callback"]}'
# Returns: {"client_id": "...", "client_secret": "..."}
```

## References

- [RFC 8414 - OAuth 2.0 Authorization Server Metadata](https://datatracker.ietf.org/doc/html/rfc8414#section-3)
- [RFC 9728 - OAuth 2.0 Protected Resource Metadata](https://datatracker.ietf.org/doc/html/rfc9728)

### PR DESCRIPTION
**Pull request by roygabriel**
## Description

MCP servers deployed at sub-paths (e.g., `https://server.smithery.ai/googledrive`) fail OAuth discovery because the client strips the path component from the server URL when constructing the OAuth base URL. This causes well-known metadata requests to hit wrong endpoints and DCR to fail with 404s.

Three bugs fixed:

1. **Path stripping** — `streamable_http.go` and `sse.go` discard the URL path when setting the OAuth base URL, losing the sub-path entirely.
2. **Wrong well-known URL construction**  `oauth.go` appends `.well-known/` to the end of the base URL instead of inserting it between host and path per RFC 8414 Section 3.
3. **No WWW-Authenticate parsing** 401 handlers return `OAuthAuthorizationRequiredError` without parsing the `resource_metadata` hint from the `WWW-Authenticate` header per RFC 9728 Section 5.1.

Fixes #697

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [x] MCP spec compatibility implementation
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Code refactoring (no functional changes)
- [ ] Performance improvement
- [ ] Tests only (no functional changes)
- [ ] Other (please describe):

## Checklist

- [x] My code follows the code style of this project
- [x] I have performed a self-review of my own code
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the documentation accordingly

## MCP Spec Compliance

- [x] This PR implements a feature defined in the MCP specification
- [x] Link to relevant spec section: [RFC 8414 Section 3 - Obtaining Authorization Server Metadata](https://datatracker.ietf.org/doc/html/rfc8414#section-3)
- [x] Link to relevant spec section: [RFC 9728 Section 5.1 - WWW-Authenticate](https://datatracker.ietf.org/doc/html/rfc9728#section-5.1)
- [x] Implementation follows the specification exactly

## Additional Information

**Files changed:**

| File | Changes |
|------|---------|
| `client/transport/oauth.go` | `ProtectedResourceMetadataURL` config field, `buildWellKnownURL` helper (RFC 8414), `extractResourceMetadataURL` helper (RFC 9728), `SetProtectedResourceMetadataURL` method, updated `getServerMetadata` discovery flow |
| `client/transport/streamable_http.go` | Include path in base URL, parse WWW-Authenticate at 401 sites |
| `client/transport/sse.go` | Include path in base URL, parse WWW-Authenticate at 401 sites |
| `client/transport/oauth_test.go` | 5 new tests (12 table-driven cases) |
| `client/transport/streamable_http_oauth_test.go` | 2 new tests |
| `client/transport/sse_oauth_test.go` | 2 new tests |

**Go version:** 1.25.7 linux/amd64

**Verification:** All existing tests pass, `go vet` clean, `golangci-lint` reports 0 issues.

**Known gap:** RFC 9728 Section 3.3 states that clients should validate retrieved protected resource metadata (e.g., confirming the `resource` field matches the server URL). This PR does not add that validation. The existing codebase did not perform it either, so this is not a regression. It could be addressed in a follow-up.

## Summary by CodeRabbit

* **New Features**
  * Allow explicit config and runtime discovery/storage of protected resource metadata URLs for OAuth
  * Extract metadata URLs from WWW-Authenticate challenges
  * Construct well-known discovery endpoints that respect server sub-paths

* **Bug Fixes**
  * Improved discovery priority, fallback logic, and error propagation during metadata retrieval and token flows
  * OAuth flows now update metadata guidance on 401 responses

* **Tests**
  * Expanded coverage for URL construction, header parsing, discovery fallbacks, and token refresh edge cases

### CODE REVIEW

File: client/transport/oauth.go [hunk 1]
```diff
@@ -310,6 +319,57 @@ func (h *OAuthHandler) SetBaseURL(baseURL string) {
 	h.baseURL = baseURL
 }
 
+// SetProtectedResourceMetadataURL stores a PRM URL discovered at runtime
+// (e.g. from the WWW-Authenticate header in a 401 response per RFC 9728).
+func (h *OAuthHandler) SetProtectedResourceMetadataURL(metadataURL string) {
+	h.protectedResourceMetadataURL = metadataURL
+}
```
[the remark this answers is not in the document]
Comment (roygabriel) [reply client/transport/oauth.go:329]: Good catch. Fixed in c3d52f2. Both the write in `SetProtectedResourceMetadataURL` and the read in `getServerMetadata` are now guarded by the existing `mu` RWMutex. Tests pass with `-race`.

**Comment by roygabriel** [comment]
Fixed in 8f11e4a. `extractResourceMetadataURL` now lower-cases the header before matching the param name (RFC 7235 case-insensitivity) while preserving the original URL casing in the extracted value. Added a test case for mixed-case param names.

File: client/transport/oauth.go [hunk 2]
```diff
@@ -427,22 +514,32 @@ func (h *OAuthHandler) getServerMetadata(ctx context.Context) (*AuthServerMetada
...
 		// If both discovery methods fail, use default endpoints based on the authorization server URL
-		metadata, err := h.getDefaultEndpoints(authServerURL)
-		if err != nil {
-			h.metadataFetchErr = fmt.Errorf("failed to get default endpoints: %w", err)
+		metadata, defErr := h.getDefaultEndpoints(authServerURL)
+		if defErr != nil {
+			h.metadataFetchErr = fmt.Errorf("failed to get default endpoints: %w", defErr)
 			return
 		}
 		h.serverMetadata = metadata
```
[the remark this answers is not in the document]
Comment (ezynda3) [reply client/transport/oauth.go:545]: @roygabriel check please
